vavr/src/main/java/io/vavr/collection/Traversable.java

```diff
@@ -190,47 +183,101 @@
     }
 
     /**
-     * Calculates the average of this elements. Returns {@code None} if this is empty, otherwise {@code Some(average)}.
-     * Supported component types are {@code Byte}, {@code Double}, {@code Float}, {@code Integer}, {@code Long},
-     * {@code Short}, {@code BigInteger} and {@code BigDecimal}.
+     * Calculates the average of this elements, assuming that the element type is {@link Number}.
+     * I.e. {@code elements.average()} is equivalent to the unsafe call:
+     *
+     * <pre>{@code
+     * elements.average(t -> ((Number) t).doubleValue())
+     * }</pre>
+     *
+     * Since we do not know if the component type {@code T} is of type {@code Number}, the
+     * {@code average()} call might throw at runtime (see examples below).
+     * <p>
+     * Please consider to use {@link #average(ToDoubleFunction)} instead.
      * <p>
      * Examples:
+     * 
      * <pre>
      * <code>
-     * List.empty().average()              // = None
-     * List.of(1, 2, 3).average()          // = Some(2.0)
-     * List.of(0.1, 0.2, 0.3).average()    // = Some(0.2)
-     * List.of("apple", "pear").average()  // throws
+     * List.empty().average()                       // = None
+     * List.of(1, 2, 3).average()                   // = Some(2.0)
+     * List.of(1.0, 10e100, 2.0, -10e100).average() // = Some(0.75)
+     * List.of("apple", "pear").average()           // throws
+     * List.of(1.0, Double.NaN).average()           // NaN
      * </code>
      * </pre>
      *
+     * Please note that Java's {@link DoubleStream#average()} uses the
+     * <a href="https://en.wikipedia.org/wiki/Kahan_summation_algorithm">Kahan summation algorithm</a>
+     * (also known as compensated summation), which has known limitations.
+     * <p>
+     * Vavr uses Neumaier's modification of the Kahan algorithm, which yields better results.
+     *
+     * <pre>{@code
+     * // = OptionalDouble(0.0) (wrong)
+     * j.u.s.DoubleStream.of(1.0, 10e100, 2.0, -10e100).average()
+     *
+     * // = Some(0.75) (correct)
+     * List.of(1.0, 10e100, 2.0, -10e100).average()
+     * }</pre>
+     *
      * @return {@code Some(average)} or {@code None}, if there are no elements
-     * @throws UnsupportedOperationException if this elements are not numeric
+     * @throws ClassCastException if this elements are not numeric
```

Comment:
@chb0github I followed you suggestions: throwing ClassCastException and adding a safe `average` variant (see below.

Reply:
I am honored

Reply:
I target this change for 1.0.0 (because it affects backward compatibility). The upcoming 0.9.1 (hopefully this weekend) will have the old exception type.